libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h

```diff
@@ -0,0 +1,204 @@
+    };
+
+    // Command IDs for CRSF_FRAMETYPE_COMMAND
+    enum CommandID {
+        CRSF_COMMAND_FC = 0x01,
+        CRSF_COMMAND_BLUETOOTH = 0x03,
+        CRSF_COMMAND_OSD = 0x05,
+        CRSF_COMMAND_VTX = 0x08,
+        CRSF_COMMAND_LED = 0x09,
+        CRSF_COMMAND_FW_UPDATE = 0x0A,
+        CRSF_COMMAND_RX = 0x10,
+    };
+
+    // Commands for CRSF_COMMAND_FC
+    enum CommandFC {
+        CRSF_COMMAND_FC_DISARM = 0x01,
+        CRSF_COMMAND_SCALE_CHANNEL = 0x02,
+    };
+
+    // Commands for CRSF_COMMAND_BLUETOOTH
+    enum CommandBluetooth {
+        CRSF_COMMAND_BLUETOOTH_RESET = 0x01,
+        CRSF_COMMAND_BLUETOOTH_ENABLE = 0x02,
+        CRSF_COMMAND_BLUETOOTH_ECHO = 0x64,
+    };
+
+    // Commands for CRSF_COMMAND_OSD
+    enum CommandOSD {
+        CRSF_COMMAND_OSD_SEND = 0x01,
+    };
+
+    // Commands for CRSF_COMMAND_VTX
+    enum CommandVTX {
+        CRSF_COMMAND_VTX_CHANNEL = 0x01,
+        CRSF_COMMAND_VTX_FREQ = 0x02,
+        CRSF_COMMAND_VTX_POWER = 0x03,
+        CRSF_COMMAND_VTX_PITMODE = 0x04,
+        CRSF_COMMAND_VTX_PITMODE_POWERUP = 0x05,
+        CRSF_COMMAND_VTX_POWER_DBM = 0x08,
+    };
+
+    // Commands for CRSF_COMMAND_LED
+    enum CommandLED {
+        CRSF_COMMAND_LED_SET_DEFAULT = 0x01,
+        CRSF_COMMAND_LED_COLOR = 0x02,
+        CRSF_COMMAND_LED_PULSE = 0x03,
+        CRSF_COMMAND_LED_BLINK = 0x04,
+        CRSF_COMMAND_LED_SHIFT = 0x05,
+    };
+
+    // Commands for CRSF_COMMAND_FW_UPDATE
+    enum CommandFirmwareUpdate {
+        CRSF_COMMAND_FIRMWARE_UPDATE_BOOTLOADER = 0x0A,
+        CRSF_COMMAND_FIRMWARE_UPDATE_ERASE = 0x0B,
+    };
+
+    // Commands for CRSF_COMMAND_RX
+    enum CommandRX {
+        CRSF_COMMAND_RX_BIND = 0x01,
+    };
+
+    enum DeviceAddress {
+        CRSF_ADDRESS_BROADCAST = 0x00,
+        CRSF_ADDRESS_USB = 0x10,
+        CRSF_ADDRESS_TBS_CORE_PNP_PRO = 0x80,
+        CRSF_ADDRESS_RESERVED1 = 0x8A,
+        CRSF_ADDRESS_PNP_PRO_CURRENT_SENSOR = 0xC0,
+        CRSF_ADDRESS_PNP_PRO_GPS = 0xC2,
+        CRSF_ADDRESS_TBS_BLACKBOX = 0xC4,
+        CRSF_ADDRESS_FLIGHT_CONTROLLER = 0xC8,
+        CRSF_ADDRESS_RESERVED2 = 0xCA,
+        CRSF_ADDRESS_RACE_TAG = 0xCC,
+        CRSF_ADDRESS_VTX = 0xCE,
+        CRSF_ADDRESS_RADIO_TRANSMITTER = 0xEA,
+        CRSF_ADDRESS_CRSF_RECEIVER = 0xEC,
+        CRSF_ADDRESS_CRSF_TRANSMITTER = 0xEE
+    };
+
+    enum ExtendedFrameOffset {
+        CRSF_EXTENDED_FRAME_LENGTH_OFFSET = 1,
+        CRSF_EXTENDED_FRAME_TYPE_OFFSET = 2,
+        CRSF_EXTENDED_FRAME_DESTINATION_OFFSET = 3,
+        CRSF_EXTENDED_FRAME_ORIGIN_OFFSET = 4,
+        CRSF_EXTENDED_FRAME_PAYLOAD_OFFSET = 5,
+    };
+
+    struct Frame {
+        uint8_t device_address;
+        uint8_t length;
+        uint8_t type;
+        uint8_t payload[CRSF_FRAMELEN_MAX - 3]; // +1 for crc
+    } PACKED;
+
+private:
+    struct RCChannelsPayload {
+        // 176 bits of data (11 bits per channel * 16 channels) = 22 bytes.
+        uint32_t chan0 : 11;
```

Comment:
Any thoughts on the best way? The setup in SBUS looks complicated - is the performance a big win?

Reply:
Ok did a test and my way is 8x faster than the SBUS way and 8 bytes smaller. So I'm going to go with my way. Test code attached in case it wasn't a fair test.

Reply:
Fixed